cmd/gardenlet/app/gardenlet.go

```diff
@@ -364,19 +373,29 @@ func NewGardenlet(ctx context.Context, cfg *config.GardenletConfiguration) (*Gar
 		K8sGardenCoreInformers: gardencoreinformers.NewSharedInformerFactory(k8sGardenClient.GardenCore(), 0),
 		KubeInformerFactory:    kubeinformers.NewSharedInformerFactory(k8sGardenClient.Kubernetes(), 0),
 		LeaderElection:         leaderElectionConfig,
+		CertificateManager:     certificateManager,
 	}, nil
 }
 
 // Run runs the Gardenlet. This should never exit.
 func (g *Gardenlet) Run(ctx context.Context) error {
 	controllerCtx, controllerCancel := context.WithCancel(ctx)
 
-	defer controllerCancel()
+	defer func() {
+		controllerCancel()
+		if g.CertificateManager != nil {
+			g.CertificateManager.Stop()
+		}
```

Comment:
Hmm, why does the `CertificateManager` have to be stopped explicitly? Please pass a context to the `CertificateManager` on Start and let it run until `ctx.Done()` is closed.
Also, I think the leadership election context should also be cancelled after successful certificate rotation, just as it happens when the gardenlet receives a `SIGINT`.
Maybe we should have a `rootCtx` which will be the parent for `controllerCtx` and `leaderElectionCtx` and will be cancelled by the certificate manager after successful rotation.

Reply:
I am passing the controller context now, then canceling it when the rotation is finished. So both the controller and the cert rotation are stopped.

What is the benefit of also canceling the leaderElectionCtx? I my tests, this setup works just fine.

Reply:
Hmm, sorry. The leadership context also was not cancelled before on `SIGINT`. It is actually cancelled when the controllers have been shutdown. Also, we don't set `leaderElectionConfig.ReleaseOnCancel`, so this anyways wouldn't have had the effect I was thinking about.
So this is fine now :)